aggregator/entry.go

```diff
@@ -315,170 +339,179 @@ func (e *Entry) recordLastAccessed(currTime time.Time) {
+	for _, pipeline := range sm.Pipelines {
+		storagePolicies := e.storagePolicies(pipeline.StoragePolicies)
+		for _, storagePolicy := range storagePolicies {
+			key := aggregationKey{
+				aggregationID: pipeline.AggregationID,
+				storagePolicy: storagePolicy,
+				pipeline:      pipeline.Pipeline,
+			}
+			idx := e.aggregations.index(key)
+			if idx < 0 {
+				return true
+			}
+			bs.Set(uint(idx))
 		}
 	}
-	return false
+	return bs.All(uint(len(e.aggregations)))
 }
 
-func (e *Entry) updatePoliciesWithLock(
-	typ unaggregated.Type,
-	id metricid.RawID,
-	ownsID bool,
-	hasDefaultPoliciesList bool,
-	sp policy.StagedPolicies,
-) error {
-	policies, useDefaultPolicies := sp.Policies()
-	if useDefaultPolicies {
-		policies = e.opts.DefaultPolicies()
+func (e *Entry) storagePolicies(policies []policy.StoragePolicy) []policy.StoragePolicy {
+	if !policy.IsDefaultStoragePolicies(policies) {
+		return policies
 	}
+	return e.opts.DefaultStoragePolicies()
+}
 
-	// Fast path to exit in case the policies didn't change.
-	if !e.hasPolicyChangesWithLock(policies) {
-		e.hasDefaultPoliciesList = hasDefaultPoliciesList
-		e.useDefaultPolicies = useDefaultPolicies
-		e.cutoverNanos = sp.CutoverNanos
-		e.metrics.policyUpdates.Inc(1)
-		return nil
+func (e *Entry) maybeCopyIDWithLock(metric unaggregated.MetricUnion) metricid.RawID {
+	// If we own the ID, there is no need to copy.
+	if metric.OwnsID {
+		return metric.ID
 	}
 
-	elemID := id
-	if !ownsID {
-		if len(e.aggregations) > 0 {
-			// If there are existing elements for this id, try reusing
-			// the id from the elements because those are owned by us.
-			for _, elem := range e.aggregations {
-				elemID = elem.Value.(metricElem).ID()
-				break
-			}
-		} else {
-			// Otherwise this is a new id so it is necessary to make a
-			// copy because it's not owned by us.
-			elemID = make(metricid.RawID, len(id))
-			copy(elemID, id)
+	// If there are existing elements for this id, try reusing
+	// the id from the elements because those are owned by us.
+	if len(e.aggregations) > 0 {
+		for _, aggVal := range e.aggregations {
+			return aggVal.elem.Value.(metricElem).ID()
 		}
 	}
 
-	// We should update the policies first.
-	newAggregations := make(map[policy.Policy]*list.Element, len(policies))
-	for _, p := range policies {
-		if elem, exists := e.aggregations[p]; exists {
-			newAggregations[p] = elem
-		} else {
-			aggTypes, err := e.decompressor.Decompress(p.AggregationID)
-			if err != nil {
-				return err
-			}
+	// Otherwise it is necessary to make a copy because it's not owned by us.
+	elemID := make(metricid.RawID, len(metric.ID))
+	copy(elemID, metric.ID)
+	return elemID
+}
+
+func (e *Entry) updateMetadatasWithLock(
+	metric unaggregated.MetricUnion,
+	hasDefaultMetadatas bool,
+	sm metadata.StagedMetadata,
+) error {
+	var (
+		elemID          = e.maybeCopyIDWithLock(metric)
+		newAggregations = make(aggregationValues, 0, initialAggregationCapacity)
```

Comment:
Not sure if it's necessary but since this is always locked, I think you can get away with keeping 2 aggregations in the entry: existingAggregations and newAggregations, when you need to update, do it to the newAggregations and swap the 2 in the end of the update.

Reply:
True, but given that this only happens once after rule update and should be infrequent, seems like an overkill to make that optimization now.